Add more protos to the repo following the spec

jvican opened the pull request:
This is still a work in progress. So far, this PR:

* Add endpoints for new bsp endpoints
* Add ScalaPlatform to build target
* Add scala test classes protos
* Add clarification to the docs that compile is intransitive

olafurpg commented on docs/bsp.md:
```diff
@@ -435,6 +435,8 @@ trait DependencySourcesItem {
 ### 1.6.7. Compile Request
 
 The compile build target request is sent from the client to the server to compile the given list of build targets.
+The compilation request will not compile these projects transitively.
```
Can you elaborate on this?

jvican replied:
Imagine the following scenario:

laughedelic replied:
💭

olafurpg commented:
Ping, any update on this?

olafurpg commented on src/main/protobuf/bsp.proto:
```diff
@@ -135,17 +150,43 @@ message ScalaBuildTarget {
     string scalaOrganization = 1;
     string scalaVersion = 2;
     string scalaBinaryVersion = 3;
+    ScalaPlatform platform = 4;
```
Seems off-sync with the spec. It's tedious to keep the two synchronized but I think it's worth the effort.

olafurpg commented on docs/bsp.md:
```diff
@@ -253,25 +248,37 @@ trait InitializeBuildResult {
-  def testProvider: Boolean
+  /** The languages the server supports compilation via method buildTarget/compile. */
+  def compileProvider: {
+    val languageIds: List[String]
```
s/val/def/

olafurpg commented on docs/bsp.md:
```diff
@@ -341,31 +348,31 @@ Notification:
-	/** The message type. See {@link MessageType}. */
-	def type: Int
+  /** The message type. See {@link MessageType}. */
+  def `type`: Int
```
Let's keep `type` un-backticked to avoid confusion for non-Scala readers that the backticks are not expected in JSON object keys.

olafurpg commented on docs/bsp.md:
```diff
@@ -652,32 +658,38 @@ trait TestParams {
 
-  /** The total number of milliseconds it took to run the tests.
-    * Should not include compile times. */
+  /** The total number of milliseconds it took to only run the tests. */
```
I think the previous docstring was clearer, it's good to explicitly say this should not include compile or packaging times.

olafurpg commented on docs/bsp.md:
```diff
@@ -377,16 +380,21 @@ object MessageType {
+additions.
+
+The `id` and optional `parentId` fields allow clients to structure logs in a hierarchical way (in a
+tree fashion, with dropdowns, et cetera) to ease readability.
```
s/, et cetera//

Same goes for "obviously" and "clearly", makes readers feel dumb when they don't understand

olafurpg commented on docs/bsp.md:
```diff
@@ -352,10 +352,13 @@ trait ShowMessageParams {
 
   /** The message id. */
-  def id: Long
+  def id: string
```
Should this be called `taskId` instead? And `parentTaskId`, respectively. The docstrings should have the explanation below on what the difference between "task id" and "request id" is.

olafurpg replied:
It would also be good to mention that this is is generated by the server, not the client.

olafurpg replied:
Should this be optional?

olafurpg commented on docs/bsp.md:
```diff
@@ -684,7 +725,12 @@ The server communicates during the initialize handshake whether this method is s
+  /** A sequence of build targets to test. */
   def targets: List[BuildTargetIdentifier]
+  
+  /** An optional number uniquely identifying a client request. */
```
> A unique identifier generated by the client to identify this request.
> The server may include this id as `requestId` in notifications such as `build/compileReport`

olafurpg commented on docs/bsp.md:
```diff
@@ -723,9 +776,11 @@ trait TestReport {
-number of `buildTarget/publishDiagnostics` and `build/logMessage` notifications during compilation
-before completing the response. The client is free to forward these messages to the LSP editor
-client.
+number of `build/publishDiagnostics` and `build/logMessage` notifications during compilation before
```
Include `build/compileReport` as another example

olafurpg commented on docs/bsp.md:
```diff
@@ -803,42 +858,57 @@ trait ScalacOptionsItem {
-This method can for example be used by a language server to attach a "Run test suite" button above the definition of a test suite via `textDocument/codeLens`.
-To render the code lens, the language server needs to map the fully qualified names of the test targets to the defining source file via `textDocument/definition`.
-Then, once users click on the button, the language server can pass the fully qualified name of the test class as an argument to the `buildTarget/test` request.
+The build target scalac options request is sent from the client to the server to query for the list
```
s/The build target scalac options request/The Scala test classes request/

olafurpg commented on docs/bsp.md:
```diff
@@ -803,42 +858,57 @@ trait ScalacOptionsItem {
-To render the code lens, the language server needs to map the fully qualified names of the test targets to the defining source file via `textDocument/definition`.
-Then, once users click on the button, the language server can pass the fully qualified name of the test class as an argument to the `buildTarget/test` request.
+The build target scalac options request is sent from the client to the server to query for the list
+of fully qualified names of test clases in a given list of targets. This method can for example be
```
End the paragraph at "list of targets" and start a new paragraph at "This method can for example"

jvican commented:
Just addressed all feedback, feel free to refine the introduction and motivation. We're still far from having something as convincing as I'd like them to be.
  
Did you have a look at the addition `buildTarget/run` and `buildTarget/scalaMainClasses`?

jvican commented:
@jastice Feel free to have a look at the changes, your input is appreciated :smile: I haven't yet implemented in the bloop side but I think that the sbt language extension has all we need to have better build.sbt file support in IntelliJ, could you confirm?

jastice commented:
@jvican From the description it's not clear to me: is `SbtBuildTarget` supplied for any target used by sbt? Is `parent` supposed to be the meta build of that target or the other way around?

Minimum information missing: sbt version (required for correct highlighting)
probably helpful: sbt project uri/id, configuration, project base dir. these could be used to merge targets back to intellij modules where necessary.

jvican commented:
@jastice Thanks for having a look!
  
The child (a normal build) has a reference to the parent (meta-build), and so on. The idea is that there will be an sbt build target per sbt project (including meta projects) so that IntelliJ can import them correctly.
  
You're right that the sbt version is missing. The project URI is already in the `BuildTarget`. The way the structure works is that the `BuildTarget` has a data field that contains the language specific information, in this case `SbtBuildTarget`.
  
What I don't understand is why you need configurations. Isn't the configuration always going to be `Compile`?

jastice commented:
Is the BuildTarget uri the exact project uri? how is the sbt project id included?

In bloop you have separate `name` and `name-test` targets that I combine back into a single IDEA module with compile and test sources/dependencies/etc respectively. How do you intend to do this for sbt?

jvican commented:
> Is the BuildTarget uri the exact project uri? how is the sbt project id included?
  
The build target URI will be the one of the sbt project URI, like `/data/rw/code/scala/foo/project`. What do you mean by sbt project id?
  
> In bloop you have separate name and name-test targets that I combine back into a single IDEA module with compile and test sources/dependencies/etc respectively. How do you intend to do this for sbt?
  
As there's only one configuration per sbt project, I was thinking of just having one IDEA module for the sbt build. Is that possible?

jvican commented:
This is ready for review again @olafurpg

olafurpg commented on docs/bsp.md:
```diff
@@ -864,6 +864,9 @@ trait ScalaBuildTarget {
   /** The target platform for this target */
   def platform: Int
+  
+  /** A sequence of Scala jars. */
```
Can you elaborate on what jars these are? scala-library, scala-reflect, scala-compiler?

olafurpg replied:
the `scalac` command line interface does not have a --scala-jars flag so it's not obvious to me  what this is

olafurpg commented on docs/bsp.md:
````diff
@@ -370,20 +367,36 @@ where `MessageType` is defined as follows:
+and a `TreeStructure` is defined as:
+
+```scala
+trait TreeStructure {
````
I'm not so keen on this name. Given that `TreeStructure` is only used in two messages, can we revert this commit and keep `nodeId` and `parentId` inlined?

olafurpg replied:
If we keep it, this should be included in the "base data structures" section.

olafurpg commented on docs/bsp.md:
```diff
@@ -147,6 +147,10 @@ trait BuildTarget {
 
+  /** The type of build target. Useful for an IDE to show
+    * targets with different kinds in the UI. */
+  def kind: Int
```
`App` is missing. What about projects that are for example of kind bench AND test?  I think it would be better to have `labels: List[String]` and leave it to clients and servers to come up with meaningful names.

olafurpg commented:
Whoop, accidentally closed the PR, sorry.

I am ok with keeping `BuildTarget.kind` if we make it a `List[String]`, I'm convinced it's useful information.

olafurpg commented on docs/bsp.md:
```diff
@@ -159,6 +163,13 @@ trait BuildTarget {
+  final val Library = 1
+  final val Test = 2
+  final val IntegrationTest = 3
+  final val Bench = 4
```
If we keep kind as a single element, 

- it should be a string instead of int, similarly as language IDs are strings
- we are missing `App` and in the future more categories may come up like `iOS`, `Android`

I personally lean towards turning `kind: Int` into `labels: List[String]` where we provide a builtin set of strings like we do with languages.

olafurpg replied:
After more thinking and offline discussions, I'm in favor of keeping `BuildTarget.kind: Int` but we should add `final val App = ...`

jvican commented:
Feedback addressed, thanks for your feedback @olafurpg :pray: I'm merging.
